### DISCUSSION TITLE
Rescoring CSMs (cross-linked peptides) by mokapot

### DISCUSSION THREAD

**Discussion by mostafakalhor** [wfondrie/mokapot#103] [Q&A]
Dear Team,

I hope you are doing well. I am Mostafa Kalhor, a PhD student at TUM, and I hope you can spare a moment to address a few queries I have regarding the utilization of Mokapot for the rescoring of Cross-Linked Spectrum Matches (CSMs).
Upon reviewing your source code, particularly focusing on how Mokapot handles False Discovery Rate (FDR) calculation for cross-linked peptides, I came across the following code snippet:
```python
fdr = np.divide(
        (one_decoy - two_decoy),
        cum_targets,
        out=np.ones_like(cum_targets, dtype=float),
        where=(cum_targets != 0),
    )
fdr[fdr < 0] = 0
```
I have a couple of questions:

In the output of many cross-link search engines, only one column specifies the specific CSM from either target or decoy, without individual details for peptide 'A' and 'B'. Would it be advisable for me to reuse FASTA files and perform a digestion step to obtain all possible peptides from the target?

Currently, I'm able to execute Mokapot using my 'rescore.tab' files. However, given the absence of information about peptide 'A' and 'B', Mokapot might interpret each CSM as a Peptide-Spectrum Match (PSM). In my 'rescore.tab' file, there's a 'label' column indicating if a CSM is from target or decoy. How precisely does Mokapot handle cross-linked peptides, and would it be necessary to introduce a dedicated column in the 'rescore.tab' file to indicate these features as CSM-specific rather than PSM-related?

I'm extremely grateful for your time and insights.

Best regards,

Mostafa Kalhor

**Comment by wfondrie** [comment]
Hi @mostafakalhor 👋 ,

Sorry to take so long to get back to you!

You found a hidden part of mokapot that was part of my first research project when I joined Bill Noble's lab as a postdoc. The goal was to implement add full support for cross-linked peptides to the Percolator algorithm, in this case within mokapot. Some research had already been done on how to estimate FDR for cross-linked peptides, which we were able to theoretically verify from the [xProphet paper](https://www.nature.com/articles/nmeth.2103). It turns out that in the cross-link setting, we really need to know whether one or both peptides in the CSM are a target or decoy to calculate a robust estimate of the FDR.

Unfortunately this project turned out to be problematic for a few reasons: (1) a large majority of cross-linking mass spectrometry studies contain relatively few cross-linked peptides, making Percolator-style on-the-fly model training unreliable (as an aside, that was why [I added static model support to Percolator](https://pubs.acs.org/doi/10.1021/acs.jproteome.9b00780)). (2) The nature of the CSM FDR calculation made it unstable to use for training, because there is unfortunate way to short circuit it: if a model ranks decoy-decoy CSMs highly, then everything is estimated to have a low FDR, which is obviously not how we want to optimize the equation. Alas, I was never able to make it work properly (or at least to a degree I was comfortable releasing), so it has lived on as a secret part of the mokapot repository since its inception.

I hope to one day revisit it, but I'm currently planning on removing it in our next release. 😞
